lib/cocoa/src/protocol/TCompactProtocol.m

```diff
@@ -0,0 +1,706 @@
+        [self writeI16:fieldID];
+    }
+    
+    lastFieldId = fieldID;
+}
+
+-(void)writeFieldStop
+{
+    [self writeByteDirect:TCType_STOP];
+}
+
+-(void)writeMapBeginWithKeyType:(int)keyType valueType:(int)valueType size:(int)size
+{
+    if (size == 0)
+    {
+        [self writeByteDirect:0];
+    }
+    else
+    {
+        [self writeVarint32:(uint32_t)size];
+        [self writeByteDirect:[self getCompactTypeForTType:keyType] << 4 | [self getCompactTypeForTType:valueType]];
+    }
+}
+
+-(void)writeListBeginWithElementType:(int)elementType size:(int)size
+{
+    [self writeCollectionBeginWithElementType:elementType size:size];
+}
+
+-(void)writeSetBeginWithElementType:(int)elementType size:(int)size
+{
+    [self writeCollectionBeginWithElementType:elementType size:size];
+}
+
+-(void)writeBool:(BOOL)b
+{
+    if (boolFieldId != nil && boolFieldName != nil && boolFieldType != nil)
+    {
+        // we haven't written the field header yet
+        [self writeFieldBeginInternalWithName:boolFieldName type:[boolFieldType intValue] fieldID:[boolFieldId intValue] typeOverride:b ? TCType_BOOLEAN_TRUE : TCType_BOOLEAN_FALSE];
+        
+        [boolFieldId release_stub];
+        [boolFieldName release_stub];
+        [boolFieldType release_stub];
+        
+        boolFieldId = nil;
+        boolFieldName = nil;
+        boolFieldType = nil;
+    }
+    else
+    {
+        // we're not part of a field, so just Write the value.
+        [self writeByteDirect:b ? TCType_BOOLEAN_TRUE : TCType_BOOLEAN_FALSE];
+    }
+}
+
+-(void)writeByte:(uint8_t)value
+{
+    [self writeByteDirect:value];
+}
+
+-(void)writeI16:(short)value
+{
+    [self writeVarint32:[self i32ToZigZag:value]];
+}
+
+-(void)writeI32:(int32_t)value
+{
+    [self writeVarint32:[self i32ToZigZag:value]];
+}
+
+-(void)writeI64:(int64_t)value
+{
+    [self writeVarint64:[self i64ToZigZag:value]];
+}
+
+-(void)writeDouble:(double)value
+{
+    //Safe bit-casting double->uint64
+    uint8_t data[8];
+    uint64_t bits = 0;
+    memcpy(data, &value, 8);
+    memcpy(&bits, data, 8);
+    
+    bits = OSSwapHostToLittleInt64(bits);
+    [mTransport write:(uint8_t*)&bits offset:0 length:8];
+}
+
+-(void)writeString:(NSString*)value
+{
+    [self writeBinary:[value dataUsingEncoding:NSUTF8StringEncoding]];
+}
+
+-(void)writeBinary:(NSData*)data
+{
+    [self writeVarint32:(uint32_t)data.length];
+    [mTransport write:data.bytes offset:0 length:data.length];
+}
+
+-(void)writeMessageEnd {}
+-(void)writeMapEnd {}
+-(void)writeListEnd {}
+-(void)writeSetEnd {}
+-(void)writeFieldEnd {}
+
+-(void)writeCollectionBeginWithElementType:(int)elementType size:(int)size
+{
+    if (size <= 14)
+    {
+        [self writeByteDirect:size << 4 | [self getCompactTypeForTType:elementType]];
+    }
+    else
+    {
+        [self writeByteDirect:0xf0 | [self getCompactTypeForTType:elementType]];
+        [self writeVarint32:(uint32_t)size];
+    }
+}
+
+-(void)writeVarint64:(UInt64)n
```

Comment:
We should consistently use either the Carbon-style `UInt64` or the C-style `uint_64_t` throughout here. Currently it's a mix - the method body uses `uint8_t` 2 lines below here, for example.